Azure Devops integrated with Permission Framework (#22591)
backstage/backstage: plugins/azure-devops/README.md

+ if ( isPermission(request.permission, azureDevOpsPullRequestReadPermission) ||
+      isPermission(request.permission, azureDevOpsPipelineReadPermission) ||
+      isPermission(request.permission, azureDevOpsGitTagReadPermission) ||
+      isPermission(request.permission, azureDevOpsReadmeReadPermission)) {
+    return createCatalogConditionalDecision(
+      request.permission,
+      catalogConditions.isEntityOwner({
+          claims: user?.identity.ownershipEntityRefs ?? [],
+       }),
+    );
+  }

+ if ( isPermission(request.permission, azureDevOpsPullRequestDashboardReadPermission) {
+ return {
+   result: AuthorizeResult.ALLOW,
+  };
+ }

  return {
    result: AuthorizeResult.ALLOW,
  };
}
```

#### Configure Azure DevOps PullRequest Dashboard Permission

To Integrate Azure DevOps PullRequest Dashboard permssion, add the following in `packages/app/src/App.tsx`

```diff
...
+ import { RequirePermission } from '@backstage/plugin-permission-react';
+ import { azureDevOpsPullRequestDashboardReadPermission } from '@backstage/plugin-azure-devops-common';
...
...
  const routes =(
    <FlatRoutes>
    ...
+   <Route
+     path="/pullrequest-dashboard"
+     element={

> +      <RequirePermission permission={azureDevOpsPullRequestDashboardReadPermission}>

Reviewer: I think this should be in the code, this saves the Adopter from having to add this and on our end we won't have to maintain this documentation.

Author: Yes I agree. I will update the `PullRequestPage` with `RequirePermission`

- You'll need to add the `EntitySwitch.Case` above from step 2 to all the entity sections you want to see Readme in. For example if you wanted to see Readme when looking at Website entities then you would need to add this to the `websiteEntityPage` section.
- The `if` prop is optional on the `EntitySwitch.Case`, you can remove it if you always want to see the tab even if the entity being viewed does not have the needed annotation
- The `maxHeight` property on the `EntityAzureReadmeCard` will set the maximum screen size you would like to see, if not set it will default to 100%

## Permission Framework

Azure DevOps plugin is now supporting the permission framework for PRs, GitTags, Pipelines & ReadMe.

> To enable the permission on the legacy backend system add the below config in `packages/backend/src/plugins/azure-devops.ts`.

Reviewer: Looks like we also need to add the common package:

```
# From your Backstage root directory
yarn add --cwd packages/backend @backstage/plugin-azure-devops-common
```

Author: `@backstage/plugin-azure-devops-common`  is already part of `@backstage/plugin-azure-devops-backend` dependencies

https://github.com/deepan10/backstage/blob/41b0f2cd3be889ce079ad8bd996577889f08db70/plugins/azure-devops-backend/package.json#L37-L39

I believe the adopters no need to add `@backstage/plugin-azure-devops-common` explictly in `packages/backend`

Reviewer: Yes, it needs to be added. If it's not there they will get an error when trying to import the permissions you added. When I did this last review I stepped though setting up the policy myself to be able to test and ran into this problem. It's always good to manually test everything.

```diff
export default async function createPlugin(
  env: PluginEnvironment,
): Promise<Router> {
  return createRouter({
    logger: env.logger,
    config: env.config,
    reader: env.reader,
+   permissions: env.permissions,
  });
}
```

### Configure Permission

To apply the permission rule add the following in `packages/backend/src/plugins/permissions.ts`.

> > Note this is example only `azureDevOpsPullRequestDashboardReadPermission` is Basic permission rest all Resource Permission. As an adopter you can configure how you wanted.

Reviewer: ```suggestion
> Note: the following is just an example of how you might want to setup permissions, as an Adopter you can configure this to fit your needs. Also all the permissions are Resource Permissions as they work with an Entity with the exception of `azureDevOpsPullRequestDashboardReadPermission`.
```

```diff
+ import {
+  azureDevOpsPullRequestReadPermission,
+  azureDevOpsPipelineReadPermission,
+  azureDevOpsGitTagReadPermission,
+  azureDevOpsReadmeReadPermission,

> +  azureDevOpsPullRequestDashboardReadPermission } from '@backstage/azure-devops-common';

Reviewer: ```suggestion
+  azureDevOpsPullRequestDashboardReadPermission } from '@backstage/plugin-azure-devops-common';
```

Reviewer: There are more imports needed to make this work. I need to add:

`isPermission` from `@backstage/plugin-permission-common`

also:

```
import {
  catalogConditions,
  createCatalogConditionalDecision,
} from '@backstage/plugin-catalog-backend/alpha';
```

Author: Instead of adding the other imports, will add like below on the top the `Permissions` section. WDYT?

> This has the prerequisite to integrate Permission Framework [Getting-started](https://backstage.io/docs/permissions/getting-started)

Reviewer: I'd prefer to have as close to a working example as possible, what I found is most people are happy with the defaults we provide in these examples. This also helps reduce the number of questions that will come up about this either in Issue or on Discord. 👍

...
async handle(
  request: PolicyQuery,
  user?: BackstageIdentityResponse,
): Promise<PolicyDecision> {
+ if ( isPermission(request.permission, azureDevOpsPullRequestReadPermission) ||
+      isPermission(request.permission, azureDevOpsPipelineReadPermission) ||
+      isPermission(request.permission, azureDevOpsGitTagReadPermission) ||
+      isPermission(request.permission, azureDevOpsReadmeReadPermission)) {

> +    return createTodoListConditionalDecision(

Reviewer: ```suggestion
+    return createCatalogConditionalDecision(
+        request.permission,
+       catalogConditions.isEntityOwner({
+          claims: user?.identity.ownershipEntityRefs ?? [],
+        }),
+      );
```

+      request.permission,
+      catalogConditions.isEntityOwner({
+          claims: user?.identity.ownershipEntityRefs ?? [],
+       }),
+    );
+  }

> + if ( isPermission(request.permission, azureDevOpsPullRequestDashboardReadPermission) {
> +   result: AuthorizeResult.ALLOW,
> + }

Reviewer: ```suggestion
+ if (
+       isPermission(
+         request.permission,
+         azureDevOpsPullRequestDashboardReadPermission,
+       )
+     ) {
+       return {
+         result: AuthorizeResult.ALLOW,
+       };
+     }
```
